Recompile when include files change in non-default directories

suexcxine opened the pull request:
Fix when app-less include files are introduced by erl_opts.

ferd commented:
There are currently passing tests for this feature which cover both include files in src/ and include/, such as https://github.com/erlang/rebar3/blob/5932c9608fc90e2cda7ff28c198c7e0b915a43d8/test/rebar_compile_SUITE.erl#L680-L753

Can you describe what yours does differently and possibly add a test for it? I'm not quite sure if this PR implies that the current code is broken (and therefore so are the tests), or if it's another use case that is being checked.

ferd commented:
Note that this patch also fails for R16B and 17.x, possibly because the `erl_anno` module was only added to Erlang in 18.x, and therefore can't be used for compiling yet.

suexcxine commented:
I place hrl files in "include/" directory as the following:
rebar.config
```
{erl_opts, [{i, "include/"}]}.
```
the include directory is in the same level as apps directory.

ferd commented:
```
λ /tmp → rebar3 new lib chkincl
...
λ /tmp → cd chkincl
λ chkincl → mkdir include
λ chkincl → echo "-define(MACRO, included).\n" > include/incl.hrl
λ chkincl → echo '-module(chk). -compile(export_all). -include("incl.hrl"). incl() -> ?MACRO. ' > src/chk.erl
λ chkincl → rebar3 shell
...
Eshell V9.0  (abort with ^G)
1> chk:incl().
included
2> q().
ok
λ chkincl → echo "-define(MACRO, modified).\n" > include/incl.hrl
λ chkincl → rebar3 shell
...
Eshell V9.0  (abort with ^G)
1> chk:incl().
modified
```
This works for me right now, without the need to even set `erl_opts` nor without the current patch.

suexcxine commented:
Your example's directory structure is "src include", namely, "src" and "include" are on the same level. But my structure is:

apps include rebar.config

ferd commented:
Oh that's why. The includes have to belong to an OTP application usually. The paths are handled relative to each app.

ferd commented:
You should get things to work the way you want without patching rebar3 in one of many ways:

1. If only one app requires the include, put the include file in `apps/<appname>/include/` and include it with `-include("file.hrl").`
2. If many apps require it, make a new app that contains the `apps/<whatever>/include/` file. Make the apps that need the include file depend on `<whatever>` in their .app.src and use `-include_lib("<whatever>/include/file.hrl").`
3. Keep the current structure but use a relative path from the apps' point of view: `-include("../../include/file.hrl").` so that it's obvious the current file structure is not portable to other layouts.

I believe those are all possibly cleaner than starting to support app-less include files.

ferd commented:
Now that I think of it, the third one should probably be equivalent to yours since erl_opts is acting there, but it's getting late here and I'll have to take another look later this week.

suexcxine commented:
Thanks for you advice. I'll try to adjust my project's structure.

tsloughter commented:
I'd like for this to get included but I think it needs some changes.

tsloughter commented:
Like I feel it should already be handled: https://github.com/erlang/rebar3/blob/master/src/rebar_erlc_compiler.erl#L492

Actually, it may just require modifying this `include_abs_dirs` function to include the ones defined in `erl_opts` https://github.com/erlang/rebar3/blob/master/src/rebar_erlc_compiler.erl#L195

suexcxine commented:
I tried to change my erl_opts in rebar.config to 
```
{erl_opts, [{i, "/Users/name/project_name/include/"}]}.
```
and the problem is gone.

@tsloughter I changed my PR according to your hint.
and changed erl_opts back to
``` 
{erl_opts, [{i, "include/"}]}.
```
and everything is ok now.

tsloughter commented:
Cool. It'll need a test before it can be merged.

suexcxine commented:
Test added.

ferd commented:
I've approved of the changes, but just for due process, have you validated that your test case failed before your patch? If not I can double-check it just to be safe.

suexcxine commented:
I checked that test case failed before this patch:
```
%%% rebar_compile_SUITE ==> recompile_when_opts_included_hrl_changes: FAILED
%%% rebar_compile_SUITE ==> {{assert,[{module,rebar_compile_SUITE},
          {line,802},
          {expression,"ModTime =/= NewModTime"},
          {expected,true},
          {value,false}]},
 [{rebar_compile_SUITE,'-recompile_when_opts_included_hrl_changes/1-fun-0-',2,
                       [{file,"/Users/chenduo/study/rebar3/_build/test/lib/rebar/test/rebar_compile_SUITE.erl"},
                        {line,802}]},
  {test_server,ts_tc,3,[{file,"test_server.erl"},{line,1529}]},
  {test_server,run_test_case_eval1,6,[{file,"test_server.erl"},{line,1045}]},
  {test_server,run_test_case_eval,9,[{file,"test_server.erl"},{line,977}]}]}
```

suexcxine commented:
But I have no idea why CI build failed:
```
%%% rebar_pkg_alias_SUITE ==> init_per_suite: FAILED
%%% rebar_pkg_alias_SUITE ==> {{badmatch,{<<"5F9AEE6E44D65632562A52FEBB27D4CD1DBF9A0431294907EBC1C51088C5D4C4">>,
            "1c8b025391e4df936985fbf5575c3bd9"}},
 [{rebar_pkg_alias_SUITE,mock_config,2,
                         [{file,"/Users/travis/build/erlang/rebar3/_build/test/lib/rebar/test/rebar_pkg_alias_SUITE.erl"},
                          {line,176}]},
  {test_server,ts_tc,3,[{file,"test_server.erl"},{line,1529}]},
  {test_server,run_test_case_eval1,6,[{file,"test_server.erl"},{line,1130}]},
  {test_server,run_test_case_eval,9,[{file,"test_server.erl"},{line,977}]}]}
```

ferd commented:
That's fine, that's some real shaky test I'm not sure why it fails on OSX from time to time, unrelated to your patch. I'll merge once I get the CI to shut up.
